### ISSUE TITLE
IME support in popup editor of any element in a diagram

### ISSUE THREAD

**Issue by pinxue** [gaphor/gaphor#2647]
### Describe the bug
So far generaleditors.py connects key-pressed event to popdown on enter key pressing, but give no chance for IME to do its work.

### To Reproduce

1. Double click any elment to get popup editor and try to input with OS input method activated, IME doesn't work.

### Expected behavior
IME should get the key events and convert input keys.

### OS
- [ ] Linux (Please put in notes the specific distro)
- [x] macOS
- [ ] Windows

Just tested on macOS, should impacts all platforms.

### Version
Version of Gaphor: latest.

### Additional information

The problem has root cause in on_enter function at gaphor/diagram/general/generaleditors.py Line 38. 

```python
     text_view.im_context_filter_keypress(text_view.get_current_event()):
```

Maybe fix it. I cannot test it yet because LibAdwaita is missing from macports.

**Comment by amolenaar** [comment]
Hi @pinxue, thank you for creating this issue.

How would IME work? Are there apps om macOS I can see it work? (I suppose I have to change my locale settings).

Does it work with gtk4-demo? I'm not sure about the state of support in GTK4 on macOS.

**Comment by pinxue** [comment]
You may switch input method here:
<img width="512" alt="image" src="https://github.com/gaphor/gaphor/assets/958237/64924def-d574-48f1-80ee-b11a1c4e2365">

If there is no IME installed, you may add input methods in Settings - Keyboard - Text Input - Input Source - Edit:
<img width="711" alt="image" src="https://github.com/gaphor/gaphor/assets/958237/6faf4a59-f75a-4cc6-9150-9a8be85dca37">

The point is generaleditors intercepts all key pressing, filter out enter key and pass all other keys into text buffer, which just bypassed IME framework, im_context_filter_keypress just pass the event to it, if the key is consumed by IME it will return true. The code snippet post previous is not so right, I removed wrong part. I will try to do some verification later.

**Comment by danyeaw** [comment]
Thanks @pinxue for the bug report, having IME support is really important.

I can reproduce this on Windows. Next I'll try to test it on Linux.

**Comment by danyeaw** [comment]
I was not able to reproduce it on Linux, it looks like GNOME always uses ibus to pull up the IME candidate window, no matter what the app is supporting.

**Comment by danyeaw** [comment]
The proposed changes `text_view.im_context_filter_keypress(text_view.get_current_event())` don't seem to fix this for me. It isn't really clear to me how the filter_keypress is supposed to work with ControllerKey.

**Comment by danyeaw** [comment]
This minimum example does work with a Gtk.Entry fine with IME.

```
import gi

gi.require_version("Gtk", "4.0")
from gi.repository import Gtk, GLib


class EntryWindow(Gtk.ApplicationWindow):
    def __init__(self, **kargs):
        super().__init__(**kargs, title="Entry Demo")

        self.set_size_request(200, 100)

        header = Gtk.HeaderBar()
        self.set_titlebar(header)

        vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=6)
        vbox.props.margin_start = 24
        vbox.props.margin_end = 24
        vbox.props.margin_top = 24
        vbox.props.margin_bottom = 24
        self.set_child(vbox)

        self.entry = Gtk.Entry()
        self.entry.set_text("Hello World")
        vbox.append(self.entry)


def on_activate(app):
    win = EntryWindow(application=app)
    win.present()


app = Gtk.Application(application_id="com.example.App")
app.connect("activate", on_activate)

app.run(None)
```

I don't think it is related to our key press event controllers at all. If I turn them all off in the instanteditors module, the problem still exists. So why does this Gtk.Entry work but ours in a popup window doesn't update. Maybe the IME popup is stealing focus away from our popup?

**Comment by danyeaw** [comment]
Yup, I think this is an upstream bug. The popover is losing focus when the IME selection Window opens, here is a min reproducible example:

```
import gi

gi.require_version("Gtk", "4.0")
from gi.repository import Gtk, GLib


class EntryWindow(Gtk.ApplicationWindow):
    def __init__(self, **kargs):
        super().__init__(**kargs, title="Entry Demo")

        self.set_size_request(200, 100)

        header = Gtk.HeaderBar()
        self.set_titlebar(header)

        vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=6)
        vbox.props.margin_start = 24
        vbox.props.margin_end = 24
        vbox.props.margin_top = 24
        vbox.props.margin_bottom = 24
        self.set_child(vbox)

        button = Gtk.Button.new_with_label("Click Me")
        button.connect("clicked", self.on_click_me_clicked)
        vbox.append(button)

        self.entry = Gtk.Entry()
        self.entry.set_text("Hello World")
        self.popover = Gtk.Popover.new()
        self.popover.set_child(self.entry)
        vbox.append(self.popover)

    def on_click_me_clicked(self, button):
        self.popover.set_visible(True)


def on_activate(app):
    win = EntryWindow(application=app)
    win.present()


app = Gtk.Application(application_id="com.example.App")
app.connect("activate", on_activate)

app.run(None)
```

**Comment by danyeaw** [comment]
Upstream issue now at https://gitlab.gnome.org/GNOME/gtk/-/issues/6041

**Comment by danyeaw** [comment]
@all-contributors please add @pinxue for bug

**Comment by pinxue** [comment]
Glad to see you root-caused it so soon and thanks for adding me. Sorry for not verifying the theory, I got some nasty dependency issue to build on my M1 Mac even with homebrew. 

So far pycairo v1.24.0 installed by poetry references symbol cairo_get_hairline  which is introduced by cairo 1.8, brew only has 1.6 and has other trouble to build from HEAD source.  Here is the lock file generated on my env [poetry.lock.txt](https://github.com/gaphor/gaphor/files/12387727/poetry.lock.txt). 

By forcing pycairo 1.22.0 in pyproject.toml, it will pass but run into :
```
Failed to load shared library 'libgtk-4.1.dylib' referenced by the typelib: dlopen(libgtk-4.1.dylib, 0x0009): Symbol not found: _cairo_ft_font_face_create_for_pattern
  Referenced from: <6DE01E30-C883-364F-A1C4-CAE5BEC1ADB1> /opt/homebrew/Cellar/pango/1.50.14/lib/libpangocairo-1.0.0.dylib
  Expected in:     <C1707E34-ED90-34F5-8B6E-05E90798E712> /opt/homebrew/Cellar/cairo/1.16.0_5/lib/libcairo.2.dylib
```
Now, I'm quite curious about how the release package for mac is constructed.

**Comment by amolenaar** [comment]
@pinxue for macOS we take whatever is current in Homebrew and bundle it.

macOS build: https://github.com/gaphor/gaphor/blob/main/.github/workflows/full-build.yml#L144
and associating package commands: https://github.com/gaphor/gaphor/blob/main/.github/actions/macos_dmg/action.yml.

I had no trouble installing Gaphor (`poetry install`) on my Mac, with updated homebrew.

**Comment by amolenaar** [comment]
On macOS I see the IME popup appear when you edit a text entry field in, say, the property editor. Nothing appears when you edit text in a popup window. It also doesn't work with drop-down widgets, like when selecting a metaclass.

**Comment by danyeaw** [comment]
So it sounds like the issues might be slightly different between macOS and Windows.

**Comment by amolenaar** [comment]
I think it is, since both use different backend code.

**Comment by amolenaar** [comment]
Fixed on macOS by https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7041.

**Comment by danyeaw** [comment]
@amolenaar Nice one!
